amplify/backend/function/fetchDomainBalance/src/config/exchangeRates.js

```diff
@@ -0,0 +1,175 @@
+const CoinGeckoConfig = require('./coinGeckoConfig');
+const { getFiatExchangeRate } = require('../api/rest/getFiatExchangeRate')
+const { getTokensList } = require('../api/rest/getTokensList')
+const { getExchangeRate, saveExchangeRate } = require('../api/graphql/requests');
+const { getStartOfDayFor } = require('../utils');
+
+const { constants: ethersConstants } = require('ethers');
+
+
+const ADDRESS_ZERO = ethersConstants.AddressZero;
+
+const isDev = process.env.ENV === 'dev';
+
+const ExchangeRatesFactory = (() => {
+
+    const exchangeRates = {};
+
+    const saveEntryToDB = async ({ tokenId, date, marketPrice }) => {
+        const marketPriceToArrayAndSupportedCurrency = Object.keys(marketPrice).map(marketPriceCurrency => {
+            const supportedCurrencyEntry = Object.entries(currencies).find(([, value]) => value === marketPriceCurrency);
+
+            if (!supportedCurrencyEntry) {
+                return null
+            }
+
+            return {
+                currency: supportedCurrencyEntry?.[0],
+                rate: marketPrice[marketPriceCurrency]
+            };
+        }).filter(entry => !!entry) ?? [];
+
+        return saveExchangeRate({
+            tokenId,
+            date,
+            marketPrice: marketPriceToArrayAndSupportedCurrency
+        })
+    };
+
+    const getUniqueDates = (dates) => {
+        const uniqueDatesSet = new Set();
+
+        dates.forEach(date => {
+            const uniqueDate = getStartOfDayFor(date);
+            uniqueDatesSet.add(uniqueDate);
+        })
+
+        return [...uniqueDatesSet];
+    }
+
+    const getTokenIdFromAddress = async ({ tokenAddress, chainId }) => {
+        const { mappings } = await CoinGeckoConfig.getConfig();
+
+        if (tokenAddress === ADDRESS_ZERO) {
+            return mappings.networkTokens[DEFAULT_NETWORK_TOKEN.symbol];
+        }
+
+        try {
+            const chain = mappings.chains[chainId] ?? chainId;
+            const data = getTokensList();
+            const token = data.find(coinItem => coinItem?.platforms?.[chain] === tokenAddress);
+
+            return token?.id;
+        } catch {
+            return null;
+        }
+    };
+
+    const getExchangeRatesForToken = async ({
+        tokenAddress,
+        currency,
+        chainId,
+        dates
+    }) => {
+        const tokenId = await getTokenIdFromAddress({ tokenAddress, chainId })
```

Comment:
Coingecko allows to get the rate by address too. Given that we're calling this in a loop, and each iteration fetches the entire token list from coingecko, it could be optimised to reduce the number of calls.

Reply:
Indeed CoinGecko offers the possibility to retrieve the exchange rates based on the token address, but it has the limitation you need to provide the number of days for which you want to retrieve the historical data
https://docs.coingecko.com/v3.0.1/reference/contract-address-market-chart
Thus, to simplify the logic for computing the actual number of days we need and the response parsing, I decided to go with this solution.
However, if you think it makes more sense to use the other endpoint, let's do it 💪